src/script/sign.h

```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+                        throw std::ios_base::failure("Duplicate Key, input witnessScript already provided");
+                    }
+                    s >> witness_script;
+                    break;
+                }
+                case PSBT_IN_BIP32_DERIVATION:
+                {
+                    DeserializeHDKeypaths(s, key, hd_keypaths);
+                    break;
+                }
+                case PSBT_IN_SCRIPTSIG:
+                {
+                    if (!final_script_sig.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, input final scriptSig already provided");
+                    }
+                    s >> final_script_sig;
+                    break;
+                }
+                case PSBT_IN_SCRIPTWITNESS:
+                {
+                    if (!final_script_witness.IsNull()) {
+                        throw std::ios_base::failure("Duplicate Key, input final scriptWitness already provided");
+                    }
+                    UnserializeFromVector(s, final_script_witness.stack);
+                    break;
+                }
+                // Unknown stuff
+                default:
+                    if (unknown.count(key) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, key for unknown value already provided");
+                    }
+                    // Read in the value
+                    std::vector<unsigned char> val_bytes;
+                    s >> val_bytes;
+                    unknown.emplace(std::move(key), std::move(val_bytes));
+                    break;
+            }
+        }
+    }
+
+    template <typename Stream>
+    PSBTInput(deserialize_type, Stream& s) {
+        Unserialize(s);
+    }
+};
+
+/** A structure for PSBTs which contains per output information */
+struct PSBTOutput
+{
+    CScript redeem_script;
+    CScript witness_script;
+    std::map<CPubKey, std::vector<uint32_t>> hd_keypaths;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+
+    bool IsNull() const;
+    PSBTOutput() {}
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        // Write the redeem script
+        if (!redeem_script.empty()) {
+            SerializeToVector(s, PSBT_OUT_REDEEMSCRIPT);
+            s << redeem_script;
+        }
+
+        // Write the witness script
+        if (!witness_script.empty()) {
+            SerializeToVector(s, PSBT_OUT_WITNESSSCRIPT);
+            s << witness_script;
+        }
+
+        // Write any hd keypaths
+        SerializeHDKeypaths(s, hd_keypaths, PSBT_OUT_BIP32_DERIVATION);
+
+        // Write unknown things
+        for (auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        s << PSBT_SEPARATOR;
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        while(!s.empty()) {
+            // read size of key
```

Comment:
The same shorter read routine is possible here.

Reply:
Done